### PR TITLE
Update on joCoreBlock & judiVStack operations

### PR DESCRIPTION
**Pull request by ziyiyin97**
Fix the forward and adjoint operations between joCoreBlock and judiVStack.

1. judiVStack might stack different kinds of structures through computations, e.g. judiWeights, judiVector, Array. Although they are all in the same type (e.g. Float32), the overall type of the judiVStack structure might be Any, which causes "jo_check_type" problems throughout the multiplications. Thus, I fix it by checking the type of each part.

2. Some JOLI operators (like joDirac) should pertain the data structure before and after operation because it expects to be identity operator. Thus, I specifically check the operator when doing joAbstractLinearOperator*judiWeights. If the operator is dirac/N*dirac or their adjoints, the input and output should be in the same structure (judiWeights). This is kind of clumsy, but I didn't come with better ideas...

3. Slightly fix the example in examples/scripts/extended_source_lsqr.jl. There were some typos in there.

4. Add an adjoint test for joCoreBlock operator working on a stacking vector of judiVector and judiWeights.

5. Slightly fix the +/-/* operators for judiWeights with single scalar, make sure the type doesn't change. (Previously, Float64 number multiplies with Float32 judiWeights will end up with Float64 judiWeights.)

6. Overload maximum/minimum for judiWeights.

### CODE REVIEW

**Comment by ziyiyin97** [comment]
The issue here [https://github.com/slimgroup/JUDI.jl/issues/38](url) is solved in this fix type branch.

File: test/runtests.jl [hunk 1]
```diff
@@ -22,6 +22,7 @@ base = ["test_abstract_vectors.jl",
         "test_composite.jl",
         "test_judiWeights.jl",
         "test_judiWavefield.jl",
+        "test_judiVStack.jl",
```
Comment (mloubout) [inline test/runtests.jl:25]: there is already a test for it , `test_composite.jl`

File: test/test_judiVStack.jl [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# Unit test for judiVStack
+# Ziyi Yin (ziyi.yin@gatech.edu)
+# Nov 2020
+
+using JUDI.TimeModeling, LinearAlgebra, PyPlot, IterativeSolvers, JOLI, Test, Printf
```
Comment (mloubout) [inline test/test_judiVStack.jl:5]: This is a modeling test, this is not testing the unit your adding wich is `I * w` that does not require any of this, Check how the other datatypes tests are done. This file need to be removed.

File: src/TimeModeling/judiWeights.jl [hunk 3]
```diff
@@ -247,6 +255,16 @@ function norm(a::judiWeights{avDT}, p::Real=2) where avDT
...
 
+#maximum
+function maximum(a::judiWeights{avDT}) where avDT
+    return max([maximum(a.weights[i]) for i=1:a.nsrc]...)
+end
+
+#minimum
+function minimum(a::judiWeights{avDT}) where avDT
+    return min([minimum(a.weights[i]) for i=1:a.nsrc]...)
+end
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:266]: add test

File: src/TimeModeling/judiWeights.jl [hunk 4]
```diff
@@ -247,6 +255,16 @@ function norm(a::judiWeights{avDT}, p::Real=2) where avDT
     return x^(1.f0/p)
 end
 
+#maximum
+function maximum(a::judiWeights{avDT}) where avDT
+    return max([maximum(a.weights[i]) for i=1:a.nsrc]...)
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:260]: add test and oneline
```
 maximum(a::judiWeights{avDT}) where avDT =   max([maximum(a.weights[i]) for i=1:a.nsrc]...)
```

File: src/TimeModeling/judiWeights.jl [hunk 5]
```diff
@@ -192,6 +194,12 @@ function *(A::joAbstractLinearOperator{ADDT,ARDT}, v::judiWeights{avDT}) where {
     catch e
         V = A.fop(vcat([vec(v.weights[i]) for i=1:v.nsrc]...))
         jo_check_type_match(ARDT,eltype(V),join(["RDT from *(joLinearFunction,judiWeights):",A.name,typeof(A),eltype(V)]," / "))
+        if A.name == "(N*joDirac)" || A.name == "joDirac" || A.name == "adjoint((N*joDirac))" || A.name == "adjoint(joDirac)"
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:197]: No, this is bad practice, you have multiple dispatch use it

```
function *(a::joDirac, v:judiWeights) = copy(v)
```

File: src/TimeModeling/judiWeights.jl [hunk 6]
```diff
@@ -133,7 +135,7 @@ end
 function *(a::judiWeights{avDT},b::Number) where avDT
     c = deepcopy(a)
     for j=1:a.nsrc
-        c.weights[j] = c.weights[j] .* b
+        c.weights[j] = convert(Array{avDT}, c.weights[j] .* b)
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:138]: conversion are bad and expensive, why is this needed now.

File: src/TimeModeling/judiWeights.jl [hunk 7]
```diff
@@ -124,7 +126,7 @@ end
 function -(a::judiWeights{avDT},b::Number) where avDT
     c = deepcopy(a)
     for j=1:a.nsrc
-        c.weights[j] = c.weights[j] .- b
+        c.weights[j] = convert(Array{avDT}, c.weights[j] .- b)
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:129]: conversion are bad and expensive, why is this needed now.

File: src/TimeModeling/judiWeights.jl [hunk 8]
```diff
@@ -114,7 +116,7 @@ end
 function +(a::judiWeights{avDT},b::Number) where avDT
     c = deepcopy(a)
     for j=1:a.nsrc
-        c.weights[j] = c.weights[j] .+ b
+        c.weights[j] = convert(Array{avDT}, c.weights[j] .+ b)
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:119]: conversion are bad and expensive, why is this needed now, if any is needed it should be on `b` only, this makes an extra copy for the conversion.

File: src/TimeModeling/judiWeights.jl [hunk 9]
```diff
@@ -5,6 +5,8 @@
 # Authors: Philipp Witte (pwitte@eos.ubc.ca), Henryk Modzelewski (hmodzelewski@eos.ubc.ca)
 # Date: June 2019
 
+import Base.maximum
+import Base.minimum
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:9]: imports go in `JUDI.jl` not here. Thought was working out of the box with indexing defined and would be necessary,

**Review by mloubout — CHANGES_REQUESTED** [review]
Thanks for this. Left some comments.

File: test/runtests.jl (hunk 1 above)
Comment (ziyiyin97) [reply test/runtests.jl:25]: Oh I could rename it as test_joCoreBlock.jl. In "test_composite.jl", we only test the linearity and stuff but the new test I wrote is to test the forward and adjoint of the joCoreBlock operator on judiVStack structure.

**Comment by mloubout** [comment]
>The issue here

Explicitly mention it in the PR description so that it is linked and automatically closed at merge "Fixes #38"

File: test/test_judiVStack.jl (hunk 2 above)
Comment (ziyiyin97) [reply test/test_judiVStack.jl:5]: This tests joCoreBlock*judiVStack and the adjoint. I think it's necessary (since before, jo_check_type didn't pass)

File: src/TimeModeling/judiWeights.jl (hunk 8 above)
Comment (ziyiyin97) [reply src/TimeModeling/judiWeights.jl:119]: Previously, Float64 number * Float32 judiWeights could generate a Float64 judiWeights, which we want to avoid. And I agree with you, I will only do the conversion for b then!

File: test/runtests.jl (hunk 1 above)
Comment (mloubout) [reply test/runtests.jl:25]: No, just look at how the test are setup and add the extra one. 

>test I wrote is to test the forward and adjoint of the joCoreBlock operator

joCoreBlock are JOLI objects and already tested in JOLI including adjoint. You are testing the part you are adding wich is the multiplication. Just build a joCorBLock, do the multiplication and test you get what you expect.
 `joOnes` is tested in `judWeights` for example

File: test/test_judiVStack.jl (hunk 2 above)
Comment (mloubout) [reply test/test_judiVStack.jl:5]: It needs a test but this is not a good one, this is overly complicated for the unit you are testing

File: src/TimeModeling/judiWeights.jl (hunk 5 above)
Comment (ziyiyin97) [reply src/TimeModeling/judiWeights.jl:197]: I have a question about this: joDirac is not a data structure, it is actually a joMatrix type operator, which generally belongs to joAbstractLinearOperator. Thus, a::joDirac gives an error `ArgumentError: invalid type for argument a in method definition for * at /Users/francisyin/.julia/dev/JUDI/src/TimeModeling/judiWeights.jl:202`

File: test/test_judiWeights.jl [hunk 10]
```diff
@@ -91,9 +91,11 @@ ftol = 1f-6
     @test isequal(length(w_vcat), 2*length(w))
     @test isequal(w_vcat.nsrc, 2*w.nsrc)
 
-    # dot, norm, abs
+    # dot, norm, abs, max, min
     @test isapprox(norm(w), sqrt(dot(w, w)))
-    @test isapprox(abs.(w.weights[1]), abs(w).weights[1]) 
+    @test isapprox(abs.(w.weights[1]), abs(w).weights[1])
+    @test isapprox(maximum(w),max([maximum(w.weights[i]) for i=1:w.nsrc]...))
+    @test isapprox(minimum(w),max([minimum(w.weights[i]) for i=1:w.nsrc]...))
```
Comment (mloubout) [inline test/test_judiWeights.jl:98]: That's the idea but would make it even more explicit. Currently this test "the right hand side does the same as copy pasting what's in the code", which is ok, but what you actually want is "I get exactly the value I am supposed to get". For example below where I choose what's in `d_ones` then explicitly check what the norm value is.


Like put the maximum value in `w`
```
w.weights[1][1,1] = 1f10
@test maximum(w) == 1f10
```

sand same for min.

File: test/test_judiVStack.jl (hunk 2 above)
Comment (mloubout) [reply test/test_judiVStack.jl:5]: All you should need os

```
c = [w1; jw2]
A = [joDirac; joSomething]
@test A*c = [w1, joSomething*w2]
```

and that's the only thing you should have to test in `test_composite`. Every test should always be as minimal as possible. And for example every sing test in `GROUP=JUDI` does not require and therefore does not install devito

Comment (ziyiyin97) [reply test/test_judiVStack.jl]: I am still a bit confused at here. To my understanding, this pull request mainly deals with the problem where joDirac type of operators output array instead of judiWeights, which joCoreBlock operator does not like. But, for any other type of joLinearFunction, when we do multiplication (either with judiVector or Weights or even PhysicalParameters), we have to output array in default because we don't have any assumption on the output. Therefore, I don't have any idea on how to make this test now -- seems the only meaningful test is to set `A = [lambda1*joDirac; lambda2*joDirac]`and test if A*c makes sense. Any other joSomething would be expected to fail the test. Then at the end of the day, we are not really testing joCoreBlock, we are just simply testing if joDirac outputs judiWeights instead of array.

Also, is it reasonable to overload multiplications of joDirac for judiVector as well? I am not sure

File: src/TimeModeling/judiWeights.jl (hunk 5 above)
Comment (mloubout) [reply src/TimeModeling/judiWeights.jl:223]: Why is this if still there, and why this whole new function?

File: test/test_composite.jl [hunk 11]
```diff
@@ -21,11 +21,13 @@ ftol = 1f-6
...
 
     # set up judiVector fr,om array
+    n = (10, 10)   # (x,y,z) or (x,z)
     dsize = (nsrc*nrec*ns, 1)
     rec_geometry = example_rec_geometry(nsrc=nsrc, nrec=nrec)
     data = randn(Float32, ns, nrec)
     d_obs = judiVector(rec_geometry, data)
-    w0 = judiWeights(rand(10, 10); nsrc=nsrc)
+    w0 = judiWeights(randn(Float32,n);nsrc=nsrc)
+    w0.weights=[randn(Float32,n) for i = 1:nsrc]
```
Comment (mloubout) [inline test/test_composite.jl:30]: ?

File: test/test_composite.jl [hunk 12]
```diff
@@ -115,7 +117,7 @@ ftol = 1f-6
     u_scale ./= 2f0
     @test isapprox(u_scale, u)
     u_scale .= 2f0 .* u_scale .+ v_scale
-    @test isapprox(u_scale, 2f0 * u + 2f0 + v)
+    @test isapprox(u_scale, 2f0 * u + 2f0 + v,rtol=1f-7)
```
Comment (mloubout) [inline test/test_composite.jl:120]: Why the tolerance change, this was working before and should still do.

File: test/test_composite.jl (hunk 11 above)
Comment (ziyiyin97) [reply test/test_composite.jl:30]: judiWeights(rand(10, 10); nsrc=nsrc) This will generate 2 sources with exactly the same weights. So I am thinking of giving them different random weights here for a general case.

Comment (mloubout) [reply test/test_composite.jl:30]: `w0 = judiWeights([randn(Float32,n) for i = 1:nsrc];nsrc=nsrc)`

I know it sounds nitpicky, but it's good practice to try to avoid unnecessary memory allocations at all time and get used to how to do it. Currently you allocate all the `randn` then allocate new ones and replace it which is wastefull.

Comment (ziyiyin97) [reply test/test_composite.jl:30]: Also, w0 = judiWeights(rand(10, 10); nsrc=nsrc) will output w0 as a Float32 judiWeights but the weights are actually in Float64

File: test/test_composite.jl (hunk 12 above)
Comment (ziyiyin97) [reply test/test_composite.jl]: Previously this w0 = judiWeights(rand(10, 10); nsrc=nsrc) is wrong and might lead to different results (since weights are in Float64). The weights are in default converted to Float32 so it does not pass the previous test. Now I slightly change it to `@test isapprox(u_scale, 2f0 * u + (2f0 + v))` which passes the test without changing the tolerance now.

File: src/TimeModeling/judiWeights.jl (hunk 5 above)
Comment (ziyiyin97) [reply src/TimeModeling/judiWeights.jl]: `# *(joLinearOperator, judiWeights)
function *(A::joAbstractLinearOperator{ADDT,ARDT}, v::judiWeights{avDT}) where {ADDT, ARDT, avDT}
    A.name == "joDirac" && return avDT(1) .* v
    A.name == "(N*joDirac)" && return avDT(A.fop.a) .* v
    A.name == "adjoint(joDirac)" && return avDT(1) .* v
    A.name == "adjoint((N*joDirac))" && return avDT(A.fop.a) .* v
    return mulJ(A, v)
end

function mulJ(A::joAbstractLinearOperator{ADDT,ARDT}, v::judiWeights{avDT}) where {ADDT, ARDT, avDT}
    A.n == size(v,1) || throw(judiWeightsException("shape mismatch"))
    jo_check_type_match(ADDT,avDT,join(["DDT for *(joLinearFunction,judiWeights):",A.name,typeof(A),avDT]," / "))
    # Evaluate as mat-mat over the weights
    n = size(v.weights[1])
    try
        # Mul may be defined for judiWeights
        V = A.fop(v)
        jo_check_type_match(ARDT,eltype(V),join(["RDT from *(joLinearFunction,judiWeights):",A.name,typeof(A),eltype(V)]," / "))
        return V
    catch e
        V = A.fop(vcat([vec(v.weights[i]) for i=1:v.nsrc]...))
        jo_check_type_match(ARDT,eltype(V),join(["RDT from *(joLinearFunction,judiWeights):",A.name,typeof(A),eltype(V)]," / "))
        return V
    end
end
`
This is as what we've discussed -- deal with special cases (diracs) seperately and other type of joLinearOperator go in mulJ?

File: test/test_composite.jl (hunk 11 above)
Comment (ziyiyin97) [reply test/test_composite.jl]: `w0 = judiWeights([randn(Float32,n) for i = 1:nsrc];nsrc=nsrc)`
This does not work because judiWeights can't tell it's too sources then the weights will be messed up.
I updated this in the latest commit by

`    weights = Array{Array}(undef, nsrc)
    for j=1:nsrc
        weights[j] = randn(Float32, n)
    end
    w0 = judiWeights(weights)
`

Comment (ziyiyin97) [reply test/test_composite.jl]: That should not work because it does not take cell array.
`w0 = judiWeights(convertToCell([randn(Float32,n) for i = 1:nsrc]))`
This solves the problem and it's in the latest commit

Comment (mloubout) [reply test/test_composite.jl]: >That should not work because it does not take cell array.
Well that's the point that's what should be fixed, 
>This solves the problem and it's in the latest commit
This does not solve anything, it will still error if some tries it it just hides the problem.

Comment (mloubout) [reply test/test_composite.jl]: `function judiWeights(weights::Union{Array{Any,1},Array{Array{iDT, N},1}}; vDT::DataType=Float32) where {iDT, N}`

Comment (ziyiyin97) [reply test/test_composite.jl]: Updated in the latest commit for 3 ways of input of judiWeights

Array{Array,1}
Array{Any,1}
Array{Array{Float32,2},1}

And how do they convert to the correct data types respectively

File: src/TimeModeling/judiWeights.jl [hunk 13]
```diff
@@ -41,26 +43,37 @@ can also be a single (non-cell) array, in which case the weights are the same fo
...
+        (eltype(weights[i]) != vDT) && (weights[i] = convert(Array{vDT},weights[i]))
+    end
+    # length of vector
+    n = 1
+    m = prod(size(weights[1]))*nsrc
+    return judiWeights{vDT}("Extended source weights",m,n,nsrc,weights)
+end
+
+#
+function judiWeights(weights::Array{Array{iDT, N},1}; vDT::DataType=Float32) where {iDT,N}
```
Comment (mloubout) [inline src/TimeModeling/judiWeights.jl:70]: why a separate constructor? This should be with above.

Comment (ziyiyin97) [reply src/TimeModeling/judiWeights.jl:70]: This one is separated because we could check the type match very fast -- we know its type by iDT. We only have to check once (iDT == vDT) to decide whether to convert types. In Array{Array,1} and Array{Any, 1}, the only way to check type match is to loop over all shots and check individually. So I make them separate. Is this OK?
